content/blog/modern-text-features/index.Rmd

```diff
@@ -0,0 +1,381 @@
+Well, honestly it is mostly to make you appreciate the labor that went into the fact that color fonts (and by extension, emojis) now just works:
+
+```{r, eval=FALSE}
+emojis <- "👩🏾‍💻🔥📊"
+
+p <- ggplot() + 
+  geom_label(
+    aes(x = 0, y = 0, label = emojis), 
+    family = "Apple Color Emoji"
+  )
+
+preview_devices(p, "emoji")
+```
+
+```{r, echo=FALSE, results='asis'}
+get_results("emoji")
+```
+
+As one can see the failures range from not being able to render anything, to rendering in monochrome.
+Further, it appears as if the devices have trouble figuring out the dimensions of the glyphs.
+One additional wrinkle is that while Cairo on macOS is capable of rendering in monochrome, it fails to get the correct emoji.
+This is because emojis relies heavily on ligatures, and the "dark-skinned woman at a computer" emoji is actually a ligature of the "woman", "dark skin" and "computer" emojis.
+
+## Font fallback
+
+In all of the above examples we have been very mindful in setting the font-face to a font that contains all the glyphs we need.
+This is not always practical, especially when you want to mix emoji and regular text.
+It is also an absolute requirement when mixing Latin and CJK (Chinese, Japanese, and Korean) text as it is infeasible to include all CJK glyphs in a single font.
+However, we are used to things just working at the system level.
+No matter the font it seems that a glyph is always displayed in e.g. browsers and text editors.
+This is because the OS is employing font fallback, which is the act of figuring out an alternative font to use when a glyph is not present in the chosen font.
+Wouldn't it be great if we could have that in a graphic device?
+Well, now we do!
+
+```{r, eval=FALSE}
+fallback_text <- "This is English, この文は日本語です 🚀"
+
+p <- ggplot() + 
+  geom_text(aes(x = 0, y = 0, label = fallback_text), size = 2.5)
+
+preview_devices(p, "fallback")
+```
+
+```{r, echo=FALSE, results='asis'}
+get_results("fallback")
+```
+
+The bottom line is that with ragg, you now don't need to think about missing glyphs in any font you choose (unless you request a character that is not covered by any font on your system).
+
+## Where's the catch
+
+Most of what we have shown today simply works automagically and may (depending on your prior frustrations with script support in R) seem too good to be true.
+Is there any catch?
+Not really.
+systemfonts, textshaping, and ragg try to be as smart as possible about text shaping and only take additional action if required.
+Further everything is heavily cached, so the impact on performance is negligible.
+
+There is something missing though, which we haven't touched upon.
+Not all scripts are LtR or RtL.
+A few, especially Asian scripts, are top-to-bottom.
+Top-to-bottom scripts are sadly not yet supported.
+This is not due to any limitation in the underlying shaping technology, but due to limitations in the R graphic engine, which assumes horizontal text in key places of the API.
```

Comment:
Is it graphic engine or _graphics_ engine?

Reply:
graphics